packages/liquidator/test/Liquidator.js

```diff
@@ -964,6 +973,157 @@ contract("Liquidator.js", function(accounts) {
+          assert.equal((await emp.getCollateral(sponsor2)).rawValue, convert("150"));
+
+          // If sponsor1 requests a withdrawal of any amount of collateral above 5 units at the given price of 1 usd per token
+          // their remaining position becomes undercollateralized. Say they request to withdraw 10 units of collateral.
+          // This places their position with a CR of: 115 / (100 * 1) * 100 = 115%. This is below the CR threshold.
+          await emp.requestWithdrawal({ rawValue: convert("10") }, { from: sponsor1 });
+
+          priceFeedMock.setCurrentPrice("1");
+          await liquidator.update();
+          await liquidator.liquidatePositions();
+          assert.equal(spy.callCount, 1); // There should be one log from the liquidation event of the withdrawal.
+
+          // There should be exactly one liquidation in sponsor1's account. The liquidated collateral should be the original
+          // amount of collateral minus the collateral withdrawn. 125 - 10 = 115
+          let liquidationObject = (await emp.getLiquidations(sponsor1))[0];
+          assert.equal(liquidationObject.sponsor, sponsor1);
+          assert.equal(liquidationObject.liquidator, liquidatorBot);
+          assert.equal(liquidationObject.state, LiquidationStatesEnum.PRE_DISPUTE);
+          assert.equal(liquidationObject.liquidatedCollateral, convert("115"));
+          assert.equal(liquidationObject.lockedCollateral, convert("125"));
+
+          // Advance the timer to the liquidation expiry.
+          const liquidationTime = liquidationObject.liquidationTime;
+          const liquidationLiveness = 1000;
+          await emp.setCurrentTime(Number(liquidationTime) + liquidationLiveness);
+
+          // Now that the liquidation has expired, the liquidator can withdraw rewards.
+          const collateralPreWithdraw = await collateralToken.balanceOf(liquidatorBot);
+          await liquidator.update();
+          await liquidator.withdrawRewards();
+          assert.equal(spy.callCount, 2); // 1 new info level events should be sent at the conclusion of the withdrawal. total 2.
+
+          // Liquidator should have their collateral increased by Sponsor1's collateral.
+          const collateralPostWithdraw = await collateralToken.balanceOf(liquidatorBot);
+          assert.equal(
+            toBN(collateralPreWithdraw)
+              .add(toBN(convert("125")))
+              .toString(),
+            collateralPostWithdraw.toString()
+          );
+
+          // Liquidation data should have been deleted.
+          assert.deepStrictEqual((await emp.getLiquidations(sponsor1))[0].state, LiquidationStatesEnum.UNINITIALIZED);
+
+          // The other two positions should not have any liquidations associated with them.
+          assert.deepStrictEqual(await emp.getLiquidations(sponsor2), []);
+        });
+        it("should trigger and only submit the emp minimum position", async () => {
```

Comment:
This test is useful but I feel like there should also one a test that goes through the full lifecycle of advancing block time and extending position using the `defenseActivationPercent`. Specifically, I think want to see a test that has realistic values, performs a "normal" liquidation, and then continues to extend the position size when it runs out of synths.

The `SolidityTestUtils` has a `defenseActivationPercent` function that could be useful for this kind of test. However, I have had issues running this within the Builder context but you should be fine with these bot tests as right now the bots don't use builder.

Reply:
Yes could use some more tests here. Much of the business logic is already tested separately in mocha but I don't think we can be too careful with this, so I will try to be more thorough.